src/number_theory/divisors.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson
+-/
+import data.nat.totient
+import number_theory.quadratic_reciprocity
+
+/-!
+# Divisor finsets
+
+This file defines sets of divisors of a natural number. This is particularly useful as background
+for defining Dirichlet convolution.
+
+## Main Definitions
+Let `n : ℕ`. All of the following definitions are in the `nat` namespace:
+ * `divisors n` is the `finset` of natural numbers that divide `n`.
+ * `proper_divisors n` is the `finset` of natural numbers that divide `n`, other than `n`.
+ * `divisors_antidiagonal n` is the `finset` of pairs `(x,y)` such that `x * y = n`.
+ * `perfect n` is true when the sum of `proper_divisors n` is `n`.
+
+## Implementation details
+ * All of `divisors 0`, `proper_divisors 0`, and `divisors_antidiagonal 0` are defined to be `∅`.
+
+## Tags
+divisors, perfect numbers
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+namespace nat
+variable (n : ℕ)
+
+/-- `divisors n` is the `finset` of divisors of `n`. As a special case, `divisors 0 = ∅`. -/
+def divisors : finset ℕ := finset.filter (λ x : ℕ, x ∣ n) (finset.Ico 1 (n + 1))
```

Comment:
Did you try replacing `finset.Ico 1 n` with `finset.range n`? (i.e. making `divisors 0 = {0}`). I did a quick experiment, and it doesn't look like either way causes more complications, they just arise in different places.

Reply:
Later on, it doesn't play nicely with arithmetic functions, where `card divisors`, as an arithmetic function, should map 0 to 0. However, I've ended up defining the arithmetic function a different way, so we'd need a theorem to tie them together, and we can move the 0 exception to the theorem connecting that arithmetic function to `card divisors`.

Reply:
Sorry, I didn't mean to imply that you should change everything just based on my half-tested idea :) Please feel free to choose whichever form works best for you!

Reply:
I like this one, actually.